cpp/src/parquet/encoding.cc

```diff
@@ -2226,30 +2248,69 @@ class DeltaLengthByteArrayDecoder : public DecoderImpl,
                                        MemoryPool* pool = ::arrow::default_memory_pool())
       : DecoderImpl(descr, Encoding::DELTA_LENGTH_BYTE_ARRAY),
         len_decoder_(nullptr, pool),
-        pool_(pool) {}
+        buffered_length_(AllocateBuffer(pool, 0)),
+        buffered_data_(AllocateBuffer(pool, 0)) {}
 
   void SetData(int num_values, const uint8_t* data, int len) override {
     num_values_ = num_values;
     if (len == 0) return;
-    int total_lengths_len = ::arrow::util::SafeLoadAs<int32_t>(data);
-    data += 4;
-    this->len_decoder_.SetData(num_values, data, total_lengths_len);
-    data_ = data + total_lengths_len;
-    this->len_ = len - 4 - total_lengths_len;
+
+    decoder_ = std::make_shared<::arrow::BitUtil::BitReader>(data, len);
+    len_decoder_.SetDecoder(num_values, decoder_);
+
+    int num_length = len_decoder_.ValidValuesCount();
+    PARQUET_THROW_NOT_OK(buffered_length_->Resize(num_length * sizeof(int32_t)));
+
+    int ret = len_decoder_.Decode(
+        reinterpret_cast<int32_t*>(buffered_length_->mutable_data()), num_length);
+    DCHECK_EQ(ret, num_length);
+    length_idx_ = 0;
+    num_valid_values_ = num_length;
+  }
+
+  void SetData(int num_values, std::shared_ptr<::arrow::BitUtil::BitReader> decoder) {
+    num_values_ = num_values;
+    decoder_ = decoder;
+
+    len_decoder_.SetDecoder(num_values, decoder_);
+
+    int num_length = len_decoder_.ValidValuesCount();
+    PARQUET_THROW_NOT_OK(buffered_length_->Resize(num_length * sizeof(int32_t)));
+
+    int ret = len_decoder_.Decode(
+        reinterpret_cast<int32_t*>(buffered_length_->mutable_data()), num_length);
+    DCHECK_EQ(ret, num_length);
+    length_idx_ = 0;
+    num_valid_values_ = num_length;
   }
 
   int Decode(ByteArray* buffer, int max_values) override {
-    using VectorT = ArrowPoolVector<int>;
-    max_values = std::min(max_values, num_values_);
-    VectorT lengths(max_values, 0, ::arrow::stl::allocator<int>(pool_));
-    len_decoder_.Decode(lengths.data(), max_values);
+    max_values = std::min(max_values, num_valid_values_);
+
+    int64_t data_size = 0;
+    const int32_t* length_ptr =
+        reinterpret_cast<const int32_t*>(buffered_length_->data()) + length_idx_;
     for (int i = 0; i < max_values; ++i) {
-      buffer[i].len = lengths[i];
-      buffer[i].ptr = data_;
-      this->data_ += lengths[i];
-      this->len_ -= lengths[i];
+      int32_t len = length_ptr[i];
+      buffer[i].len = len;
+      data_size += len;
+    }
+    length_idx_ += max_values;
+
+    PARQUET_THROW_NOT_OK(buffered_data_->Resize(data_size));
+    if (decoder_->GetBatch(8, buffered_data_->mutable_data(),
+                           static_cast<int>(data_size)) != static_cast<int>(data_size)) {
```

Comment:
It seems a bit weird to call both `len_decoder_` and `decoder_` methods.

More fundamentally, what happens if the user calls `Decode` incrementally (i.e. with `max_values < num_values_values_`)? It seems you should compute the position of the byte array data in `SetData` instead of assuming the `decoder_` is at the end of the encoded length data here.